lib/rules/no-dupe-else-if.js

```diff
@@ -0,0 +1,122 @@
+/**
+ * @fileoverview Rule to disallow duplicate conditions in if-else-if chains
+ * @author Milos Djermanovic
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const astUtils = require("./utils/ast-utils");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Determines whether the first given array is a subset of the second given array.
+ * @param {Function} comparator A function to compare two elements, should return `true` if they are equal.
+ * @param {Array} arrA The array to compare from.
+ * @param {Array} arrB The array to compare against.
+ * @returns {boolean} `true` if the array `arrA` is a subset of the array `arrB`.
+ */
+function isSubsetByComparator(comparator, arrA, arrB) {
+    return arrA.every(a => arrB.some(b => comparator(a, b)));
+}
+
+/**
+ * Splits the given node by the given logical operator.
+ * @param {string} operator Logical operator `||` or `&&`.
+ * @param {ASTNode} node The node to split.
+ * @returns {ASTNode[]} Array of conditions that makes the node when joined by the operator.
+ */
+function splitByLogicalOperator(operator, node) {
+    if (node.type === "LogicalExpression" && node.operator === operator) {
+        return [...splitByLogicalOperator(operator, node.left), ...splitByLogicalOperator(operator, node.right)];
+    }
+    return [node];
+}
+
+const splitByOr = splitByLogicalOperator.bind(null, "||");
+const splitByAnd = splitByLogicalOperator.bind(null, "&&");
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        type: "problem",
+
+        docs: {
+            description: "disallow duplicate conditions in if-else-if chains",
+            category: "Possible Errors",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/no-dupe-else-if"
+        },
+
+        schema: [],
+
+        messages: {
+            unexpected: "This branch can never execute, its condition is a duplicate or covered by previous conditions in the if-else-if chain."
```

Comment:
Thoughts on dropping the first clause?

```suggestion
            unexpected: "This condition is a duplicate or covered by previous conditions in the if-else-if chain."
```

If not, can we change it to the following?

```suggestion
            unexpected: "This branch can never execute. Its condition is a duplicate or covered by previous conditions in the if-else-if chain."
```

Reply:
I thought it might be good to emphasize the consequence that the branch can never execute (btw. is that correct grammar?) to make it more clear that the condition is fully covered by the previous conditions and that the code certainly has a bug or at least an unnecessary unreachable statement or block in if's `consequent`.

For instance, an advanced future version of this rule (or another rule) could also report redundant parts in otherwise valid conditions:

```js
if (a) {
    // ...
} else if (a || b) {
   // ...
}
```

This isn't necessarily a bug, the code might be working fine and there are no unreachable parts, but the `a || b` condition can be replaced with just `b`.

Reply:
Makes sense to me! The updated version has correct grammar, so this LGTM!